src/ripple/app/tx/impl/TxQImpl.cpp

```diff
@@ -0,0 +1,969 @@
+//------------------------------------------------------------------------------
+/*
+    This file is part of rippled: https://github.com/ripple/rippled
+    Copyright (c) 2012, 2013 Ripple Labs Inc.
+
+    Permission to use, copy, modify, and/or distribute this software for any
+    purpose  with  or without fee is hereby granted, provided that the above
+    copyright notice and this permission notice appear in all copies.
+
+    THE  SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+    WITH  REGARD  TO  THIS  SOFTWARE  INCLUDING  ALL  IMPLIED  WARRANTIES  OF
+    MERCHANTABILITY  AND  FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+    ANY  SPECIAL ,  DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+    WHATSOEVER  RESULTING  FROM  LOSS  OF USE, DATA OR PROFITS, WHETHER IN AN
+    ACTION  OF  CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+
+#include <ripple/app/tx/TxQ.h>
+#include <ripple/app/tx/apply.h>
+#include <ripple/app/tx/impl/applyImpl.h>
+#include <ripple/app/tx/impl/Transactor.h>
+#include <ripple/app/ledger/OpenLedger.h>
+#include <ripple/protocol/st.h>
+#include <ripple/protocol/Feature.h>
+#include <ripple/protocol/JsonFields.h>
+#include <boost/intrusive/set.hpp>
+#include <limits>
+
+namespace ripple {
+
+//////////////////////////////////////////////////////////////////////////
+
+static
+std::uint64_t
+getRequiredFeeLevel(TxType txType)
+{
+    if ((txType == ttAMENDMENT) || (txType == ttFEE))
+        return 0;
+
+    // For now, all valid non-pseudo transactions cost 256 fee units
+    // This code can be changed to support variable transaction fees
+    return 256;
+}
+
+static
+std::uint64_t
+getFeeLevelPaid(
+    STTx const& tx,
+    std::uint64_t baseRefLevel,
+    std::uint64_t refTxnCostDrops)
+{
+    // Compute the minimum XRP fee the transaction could pay
+    auto required_fee_units = getRequiredFeeLevel(tx.getTxnType());
+
+    if (required_fee_units == 0 ||
+        refTxnCostDrops == 0)
+        // If nothing is required, or the cost is 0,
+        // the level is effectively infinite.
+        return std::numeric_limits<std::uint64_t>::max();
+
+    // TODO: getRequiredFeeLevel(ttREFERENCE)?
+    auto reference_fee_units =
+        ripple::getRequiredFeeLevel(ttACCOUNT_SET);
+    return mulDivNoThrow(tx[sfFee].xrp().drops(),
+        baseRefLevel * reference_fee_units,
+            refTxnCostDrops * required_fee_units);
```

Comment:
In principal, yes. In practice, no.

* `baseRefLevel` at all the callers is a const 256. (https://github.com/ximinez/rippled/blob/fee/src/ripple/app/tx/impl/TxQImpl.cpp#L187).
* `refTxnCostDrops` is the base fee from the `Transactor` - it's going to be on the order of 10, though it can be higher with multisign, it just multiplies by the number of signatures, and that's restricted.
* `referenceFeeUnits` and `requiredFeeUnits` are going to be on the order of 256 as well. We can make it higher in the future to make a particular transaction more expensive to queue, but I can't see us making it `2^48` times higher.